pallene/assignment_conversion.lua

```diff
@@ -143,27 +148,69 @@ function Converter:apply_transformations()
                 { value = decl._type }
             )
             self:add_box_type(decl.loc, typ)
-            decl._type = typ
 
             local init_exp_update = self.update_init_exp_of_decl[decl]
 
             if init_exp_update then
                 local old_exp = init_exp_update.node
+                decl._type = typ
                 local update  = init_exp_update.update_fn
 
                 local new_node = ast.Exp.InitList(old_exp.loc, {{ name = "value", exp = old_exp }})
                 new_node._type = typ
                 update(new_node)
+
+            elseif self.lambda_of_param[decl] then
+                --- Function parameters that are captured are implementing by "proxy"-ing them.
+                --- Consider the following function that returns a closure:
+                --- ```
+                --- function m.foo(n: integer)
+                ---   -- capture n
+                --- end
+                --- ```
+                --- Since we cannot transform the declaration node of a function parameter,
+                --- we create a variable to represent the boxed parameter which can be captured.
+                --- ```
+                --- function m.foo(n: integer)
+                ---   local $n: $T = { value = n }
+                ---   -- capture and mutate $n
+                --- end
+                local param = ast.Exp.Var(decl.loc, ast.Var.Name(decl.loc, decl.name))
+                param.var._def = checker.Def.Variable(decl)
+                param.var._type = decl._type
+                param._type = decl._type
+                local decl_lhs = ast.Decl.Decl(decl.loc, "$"..decl.name, typ)
+                local decl_rhs = ast.Exp.InitList(decl.loc, {{ name = "value", exp = param }})
+                decl_rhs._type = typ
+                decl_lhs._type = typ
+
+                local new_node = ast.Stat.Decl(decl.loc, { decl_lhs }, { decl_rhs })
+
+                local lambda = self.lambda_of_param[decl]
+                table.insert(lambda.body.stats, 1, new_node)
```

Comment:
Two things that I'm not sure are a problem, but might be:

1. I suspect that the created statements might be in the opposite order of the parameters
2. This `table.insert` might result in O(n^2) behavior if we have many captured parameters

We might be able to avoid this by keeping the created statements in a list and then inserting them all at once, at the end of the program.

Reply:
> I suspect that the created statements might be in the opposite order of the parameters

Since we're merely referencing the parameters by name while creating the record boxes, this might not be a problem. What do you think?

Reply:
About the`table.insert`, I think the older method of doing assignment conversion at the end of each function was more suited to this. If we want to store all proxy statements in a table and `table.concat` it with `lambda.body.stats`, we would have to go over the `self.mutated_decls` lists again, and we'd have to keep track of which mutated decls are updated and which aren't. And a map that tells us which list of proxy decl stats belong to which lambda.

Reply:
The reason I'm thinking about the order of the statements, is that it can affect the order of the local variables that are created, which in turn can affect the `gc.lua`. But there's a decent chance that it doesn't really matter.

>  And a map that tells us which list of proxy decl stats belong to which lambda.

I think we would still need that map even if we used the `exit_lambda`? My gut instinct is to keep doing it without the exit_lambda. But if it really is easier to go back to exit lambda that would be ok.

Reply:
This did not take as many changes as I thought it would!